### ISSUE TITLE
Look into fixed point forms

### ISSUE THREAD

**Issue by saulshanabrook** [Quansight-Labs/uarray#35]
After reading http://conal.net/papers/essence-of-ad/ I wonder if we can convert our functional forms into fixed point. aka instead of having lambda functions turn them all into composition and other higher level forms.

Why? Not sure, does this get us AD? Need to review that paper more to understand crux of requirements.

**Comment by saulshanabrook** [comment]
I just watched [this talk on Compiling to Categories](http://podcasts.ox.ac.uk/compiling-categories), which is Conal's earlier paper that underlies this differentiation work.

## Notes (*italics* = my comments)
FYI: We are working in haskell, so some things are specific to that.


Two important concepts: overloading and functions/lamda. Overloading is important so that we can keep UX while changing implementation. However, this isn't compatible with functions. We can't overload how function application works. *In uarray we do overload function application, by staying away from python language proper and developing a new language inside of it*

So we could fix this by adding abstraction the language, but instead let's eliminate lambda expressions all together

<img width="1370" alt="screen shot 2018-11-06 at 1 00 12 pm" src="https://user-images.githubusercontent.com/1186124/48085621-cafbab80-e1c8-11e8-9ba1-94f7e5c3f352.png">

*In python, we don't have compiler plugins. So we can just turn `lambda i: i` into `identity`. We have to do AST/bytecode parsing. So I guess we do sort of have compiler plugins in that way, but it's just a pain. also, I imagine Python might be harder to write this sort of transformations in then haskell? Anyway, we can avoid all this by forcing users to leave language they know and right in higher level function specification. PyTorch get's around this by doing tracing, and we could do something similar. Unclear what best solution is here in Python land. Not sure if it matters though, if users have to be a little more abstract specifically, then that's OK for now, if they get some helpful results (like compiling to GPU/FPGA).*


So we replace functions with things we can overload, then we overload them! First example, overloading them to produce graph as representation of function. Here we should from user input, to compiled form, to graph:

<img width="1399" alt="screen shot 2018-11-06 at 1 06 26 pm" src="https://user-images.githubusercontent.com/1186124/48085840-655bef00-e1c9-11e8-90b7-338adf047527.png">

*Similar to our process. We go from user input, to underlying representation, then compile that to some output*

This is example of using haskell's built in sum operator:

<img width="1373" alt="screen shot 2018-11-06 at 1 08 36 pm" src="https://user-images.githubusercontent.com/1186124/48086568-30e93280-e1cb-11e8-96ec-4fb11b44f192.png">

*Pretty cool that you can recompile like this, taking existing implementation. Probably get that from how pure haskell is. Hard to imagine similar thing in Python (i.e. taking existing sum implementation and re-compiling it to this)*

*Here we something similar to how I have to specify the dimensionality before we compile. Here we have to specify size of array before we can compile folding*


Then we can also compile to Verilog (original goal of work):

<img width="1401" alt="screen shot 2018-11-06 at 1 09 52 pm" src="https://user-images.githubusercontent.com/1186124/48086708-7f96cc80-e1cb-11e8-9d01-1bee4261eb17.png">


And we can do differentiation:
<img width="1385" alt="screen shot 2018-11-06 at 1 13 28 pm" src="https://user-images.githubusercontent.com/1186124/48086739-8f161580-e1cb-11e8-9cf9-7bff9c9ff9d8.png">


We can also chain this with the graph output, to take some forms, differentiate them, then take graph.

*This is similar to how everything stays in matchpy land, even compilation. So we could do similar chaining if we implement differentiation*

## Conclusions

We could use these same rules to transform `Function` calls to fixed point forms.

**Comment by dhirschfeld** [comment]
AD is a huge usecase for me so I was wondering if that's something `uarray` would support directly of if that would be something pushed down into the backends?

The benefit I see of keeping it at the `uarray` level is that I could then use `dask` to execute it efficiently on our cluster. Maybe that's something that could be done in either case? It's on my list of things to investigate...

**Comment by saulshanabrook** [comment]
Yeah I see it as something `uarray` can support directly. Or to be more precise, something we could do on a tree of matchpy array expressions. It could be packaged in core `uarray` or some other package that extends it.

If you have some specific use cases you could point me to, I could also look into it. I am looking for some to drive development.

**Comment by dhirschfeld** [comment]
I'll try to come up with an example....
